### PR TITLE
Feat/waterfall support split total

### PR DESCRIPTION
**Pull request by xuefei1313**
[[中文版模板 / Chinese template](https://github.com/VisActor/VChart/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [X] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Release
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Other (about what?)

### 🐞 Bugserver case id

65defce7855a8400b5ecdc41

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

**Comment by xuefei1313** [comment]
支持瀑布图的总计能够被细分为不同分组的值

File: packages/vchart/src/data/transforms/waterfall.ts [hunk 1]
```diff
@@ -51,79 +66,217 @@ export const waterfall = (lastData: Array<Datum>, op: IWaterfallOpt) => {
...
+  let end = total.end;
+  const startIndex = totalData.length - +d[totalSpec.collectCountField];
+  const endIndex = totalData.length - 1;
+  if (startIndex < 0) {
+    console.warn('total.collectCountField error');
+  } else {
+    start = totalData[startIndex].start;
+  }
+  if (endIndex < 0) {
+    console.warn('total.collectCountField error');
```
Comment (xile611) [inline packages/vchart/src/data/transforms/waterfall.ts:261]: 应该用统一的 log函数吧

File: packages/vchart/src/data/transforms/waterfall.ts [hunk 2]
```diff
@@ -51,79 +66,217 @@ export const waterfall = (lastData: Array<Datum>, op: IWaterfallOpt) => {
...
-              if (totalSpec.valueField && !isNil(d[totalSpec.valueField])) {
-                end = precisionAdd(start, +d[totalSpec.valueField]);
-              }
-            }
-            d[startAs] = start;
-            d[endAs] = end;
-            d[valueField] = end - start;
-            total.start = start;
-            total.end = end;
+    // 1.10.0 当前的key对应的数据中有一个总计。并且还有其他的分组值时，总计的计算逻辑
```
Comment (xile611) [inline packages/vchart/src/data/transforms/waterfall.ts:69]: 这里是打算放到 1.10.0 支持吗？
感觉注释不是很明确

File: packages/vchart/src/data/transforms/waterfall.ts [hunk 3]
```diff
@@ -51,79 +66,217 @@ export const waterfall = (lastData: Array<Datum>, op: IWaterfallOpt) => {
...
+    if (indexData.length > 1) {
+      const isTotalCheck = (d: Datum) => {
+        if (!totalSpec || totalSpec.type === 'end') {
+          if (index === indexValues.length - 1) {
+            return true;
           }
+        } else if (totalSpec.type === 'field' || totalSpec.type === 'custom') {
```
Comment (xile611) [inline packages/vchart/src/data/transforms/waterfall.ts:76]: 这是新增的配置吗？
需要更新一下文档和类型定义

File: packages/vchart/src/data/transforms/waterfall.ts (hunk 1 above)
Comment (xuefei1313) [reply packages/vchart/src/data/transforms/waterfall.ts:261]: 是的，我去改一哈

File: packages/vchart/src/data/transforms/waterfall.ts (hunk 2 above)
Comment (xuefei1313) [reply packages/vchart/src/data/transforms/waterfall.ts:69]: 嗯，当时以为会在1.10.0 跟版本。现在如果提交合并的话，会在哪个版本上线呢？

File: packages/vchart/src/data/transforms/waterfall.ts (hunk 3 above)
Comment (xuefei1313) [reply packages/vchart/src/data/transforms/waterfall.ts:76]: 不是新增配置，是之前就存在的配置。

File: packages/vchart/src/data/transforms/waterfall.ts (hunk 2 above)
Comment (xile611) [reply packages/vchart/src/data/transforms/waterfall.ts]: 提交到 develop 是1.9.5发版

**Review by xile611 — APPROVED** [review]

**Pull request merged**